File: app/src/main/java/com/scatl/uestcbbs/module/search/view/SearchActivity.java
package com.scatl.uestcbbs.module.search.view;

import android.content.Intent;
import android.view.KeyEvent;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.TextView;

import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;

import com.jaeger.library.StatusBarUtil;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.base.BaseActivity;
import com.scatl.uestcbbs.callback.OnRefresh;
import com.scatl.uestcbbs.custom.MyLinearLayoutManger;
import com.scatl.uestcbbs.entity.SearchPostBean;
import com.scatl.uestcbbs.entity.SearchUserBean;
import com.scatl.uestcbbs.module.post.view.PostDetailActivity;
import com.scatl.uestcbbs.module.search.adapter.SearchPostAdapter;
import com.scatl.uestcbbs.module.search.adapter.SearchUserAdapter;
import com.scatl.uestcbbs.module.search.presenter.SearchPresenter;
import com.scatl.uestcbbs.module.user.view.UserDetailActivity;
import com.scatl.uestcbbs.util.ColorUtil;
import com.scatl.uestcbbs.util.CommonUtil;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.RefreshUtil;
import com.scatl.uestcbbs.util.SharePrefUtil;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.RefreshState;

public class SearchActivity extends BaseActivity<SearchPresenter> implements SearchView, View.OnKeyListener {

    private Toolbar toolbar;
    private RadioButton byPost, byUser;
    private AppCompatEditText keyWord;
    private Button searchBtn;
    private TextView hint;
    private SmartRefreshLayout refreshLayout;
    private RecyclerView recyclerView;
    private SearchUserAdapter searchUserAdapter;
    private SearchPostAdapter searchPostAdapter;

    private int page = 1;

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_search;
    }

    @Override
    protected void findView() {
        toolbar = findViewById(R.id.toolbar);
        byPost = findViewById(R.id.search_radio_btn_by_post);
        byUser = findViewById(R.id.search_radio_btn_by_user);
        keyWord = findViewById(R.id.search_keyword_edittext);
        refreshLayout = findViewById(R.id.search_refresh);
        recyclerView = findViewById(R.id.search_rv);
        searchBtn = findViewById(R.id.search_btn);
        hint = findViewById(R.id.search_hint);
    }

    @Override
    protected void initView() {
        super.initView();

        keyWord.setOnKeyListener(this);
        searchBtn.setOnClickListener(this);

        CommonUtil.showSoftKeyboard(this, keyWord, 1);

        searchPostAdapter = new SearchPostAdapter(R.layout.item_simple_post, SharePrefUtil.isHideAnonymousPost(this));
        searchUserAdapter = new SearchUserAdapter(R.layout.item_search_user);
        recyclerView.setLayoutManager(new MyLinearLayoutManger(this));
        recyclerView.setLayoutAnimation(AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_scale_in));
    }

    @Override
    protected SearchPresenter initPresenter() {
        return new SearchPresenter();
    }

    @Override
    protected void onClickListener(View view) {

        if (view.getId() == R.id.search_btn) {
            CommonUtil.hideSoftKeyboard(SearchActivity.this, keyWord);
            refreshLayout.autoRefresh(0, 300, 1, false);
        }
    }

    private void startSearch() {
        if (byPost.isChecked()) {
            presenter.searchPost(page, SharePrefUtil.getPageSize(SearchActivity.this), keyWord.getText().toString(), this);
        } else {
            presenter.searchUser(page, SharePrefUtil.getPageSize(SearchActivity.this),
                    keyWord.getText().toString().replaceAll(" ", "").replaceAll("\n", ""),this);
        }
    }

    @Override
    protected void setOnItemClickListener() {
        searchPostAdapter.setOnItemClickListener((adapter, view, position) -> {
            Intent intent = new Intent(SearchActivity.this, PostDetailActivity.class);
            intent.putExtra(Constant.IntentKey.TOPIC_ID, searchPostAdapter.getData().get(position).topic_id);
            startActivity(intent);
        });

        searchPostAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            Intent intent = new Intent(SearchActivity.this, UserDetailActivity.class);
            intent.putExtra(Constant.IntentKey.USER_ID, searchPostAdapter.getData().get(position).user_id);
            startActivity(intent);
        });

        searchUserAdapter.setOnItemClickListener((adapter, view, position) -> {
            Intent intent = new Intent(SearchActivity.this, UserDetailActivity.class);
            intent.putExtra(Constant.IntentKey.USER_ID, searchUserAdapter.getData().get(position).uid);
            startActivity(intent);
        });
    }

    @Override
    protected void setOnRefreshListener() {
        RefreshUtil.setOnRefreshListener(this, refreshLayout, new OnRefresh() {
            @Override
            public void onRefresh(RefreshLayout refreshLayout) {
                page = 1;
                startSearch();
            }

            @Override
            public void onLoadMore(RefreshLayout refreshLayout) {
                startSearch();
            }
        });
    }

    @Override
    public void onSearchUserSuccess(SearchUserBean searchUserBean) {
        page = page + 1;
        hint.setText("");
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            if (searchUserBean.has_next == 1) {
                refreshLayout.finishRefresh();
            } else {
                refreshLayout.finishRefreshWithNoMoreData();
            }
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            if (searchUserBean.has_next == 1) {
                refreshLayout.finishLoadMore(true);
            } else {
                refreshLayout.finishLoadMoreWithNoMoreData();
            }
        }

        if (searchUserBean.page == 1) {
            recyclerView.setAdapter(searchUserAdapter);
            recyclerView.scheduleLayoutAnimation();
            searchUserAdapter.setNewData(searchUserBean.body.list);
        } else {
            searchUserAdapter.addData(searchUserBean.body.list);
        }
        if (searchUserAdapter.getData().size() == 0) hint.setText("啊哦，没有数据");
    }

    @Override
    public void onSearchUserError(String msg) {
        hint.setText(msg);
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            refreshLayout.finishRefresh();
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            refreshLayout.finishLoadMore(false);
        }
    }

    @Override
    public void onSearchPostSuccess(SearchPostBean searchPostBean) {
        page = page + 1;
        hint.setText("");
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            if (searchPostBean.has_next == 1) {
                refreshLayout.finishRefresh();
            } else {
                refreshLayout.finishRefreshWithNoMoreData();
            }
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            if (searchPostBean.has_next == 1) {
                refreshLayout.finishLoadMore(true);
            } else {
                refreshLayout.finishLoadMoreWithNoMoreData();
            }
        }

        if (searchPostBean.page == 1) {
            recyclerView.setAdapter(searchPostAdapter);
            recyclerView.scheduleLayoutAnimation();
            searchPostAdapter.addSearchPostData(searchPostBean.list, true);
        } else {
            searchPostAdapter.addSearchPostData(searchPostBean.list, false);
        }

        if (searchPostAdapter.getData().size() == 0) hint.setText("啊哦，没有数据");
    }

    @Override
    public void onSearchPostError(String msg) {
        hint.setText(msg);
        if (refreshLayout.getState() == RefreshState.Refreshing) {
            refreshLayout.finishRefresh();
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            refreshLayout.finishLoadMore(false);
        }
    }

    @Override
    public boolean onKey(View v, int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_ENTER && event.getAction() == KeyEvent.ACTION_DOWN) {
            CommonUtil.hideSoftKeyboard(SearchActivity.this, v);
            refreshLayout.autoRefresh(0, 300, 1, false);
        }
        return false;
    }

    @Override
    protected void setStatusBar() {
        StatusBarUtil.setColor(
                this,
                ColorUtil.getAttrColor(this, R.attr.colorSurface), 0);
    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/search/view/SearchActivity.kt
package com.scatl.uestcbbs.module.search.view

import android.content.Intent
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.animation.AnimationUtils
import com.jaeger.library.StatusBarUtil
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.annotation.ToastType
import com.scatl.uestcbbs.base.BaseVBActivity
import com.scatl.uestcbbs.databinding.ActivitySearchBinding
import com.scatl.uestcbbs.entity.SearchPostBean
import com.scatl.uestcbbs.entity.SearchUserBean
import com.scatl.uestcbbs.module.post.view.NewPostDetailActivity
import com.scatl.uestcbbs.module.search.presenter.SearchPresenter
import com.scatl.uestcbbs.module.search.adapter.SearchPostAdapter
import com.scatl.uestcbbs.module.search.adapter.SearchUserAdapter
import com.scatl.uestcbbs.module.user.view.UserDetailActivity
import com.scatl.uestcbbs.util.CommonUtil
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.SharePrefUtil
import com.scatl.uestcbbs.util.showToast
import com.scatl.util.ColorUtil
import com.scwang.smart.refresh.layout.api.RefreshLayout

/**
 * Created by sca_tl at 2023/2/8 10:21
 */
class SearchActivity : BaseVBActivity<SearchPresenter, SearchView, ActivitySearchBinding>(),
    SearchView, View.OnKeyListener, TextWatcher {

    private lateinit var searchUserAdapter: SearchUserAdapter
    private lateinit var searchPostAdapter: SearchPostAdapter
    private var page = 1

    override fun getViewBinding() = ActivitySearchBinding.inflate(layoutInflater)

    override fun initPresenter() = SearchPresenter()

    override fun getContext() = this

    override fun initView(theftProof: Boolean) {
        super.initView(false)
        bindClickEvent(mBinding.edittext, mBinding.searchBtn)
        mBinding.edittext.addTextChangedListener(this)
        mBinding.radioGroup.setOnCheckedChangeListener { group, checkedId ->
            afterTextChanged(mBinding.edittext.editableText)
        }

        CommonUtil.showSoftKeyboard(this, mBinding.edittext, 1)

        searchPostAdapter = SearchPostAdapter(R.layout.item_simple_post)
        searchUserAdapter = SearchUserAdapter(R.layout.item_search_user)
        mBinding.rv.layoutAnimation = AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_scale_in)
    }

    override fun onClick(v: View) {
        super.onClick(v)
        if (v == mBinding.searchBtn) {
            CommonUtil.hideSoftKeyboard(this@SearchActivity, mBinding.edittext)
            mBinding.refreshLayout.autoRefresh(0, 300, 1f, false)
        }
    }

    override fun setOnItemClickListener() {
        searchPostAdapter.setOnItemClickListener { adapter, view, position ->
            val intent = Intent(this@SearchActivity, NewPostDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.TOPIC_ID, searchPostAdapter.data[position].topic_id)
            }
            startActivity(intent)
        }

        searchPostAdapter.setOnItemChildClickListener { adapter, view, position ->
            val intent = Intent(this@SearchActivity, UserDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.USER_ID, searchPostAdapter.data[position].user_id)
            }
            startActivity(intent)
        }

        searchUserAdapter.setOnItemClickListener { adapter, view, position ->
            val intent = Intent(this@SearchActivity, UserDetailActivity::class.java).apply {
                putExtra(Constant.IntentKey.USER_ID, searchUserAdapter.data[position].uid)
            }
            startActivity(intent)
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        page = 1
        startSearch()
    }

    override fun onLoadMore(refreshLayout: RefreshLayout) {
        startSearch()
    }

    private fun startSearch() {
        val inputText = mBinding.edittext.text.toString()
        if (mBinding.btnByPost.isChecked) {
            mPresenter?.searchPost(page, SharePrefUtil.getPageSize(this@SearchActivity), inputText)
        } else {
            mPresenter?.searchUser(page, SharePrefUtil.getPageSize(this@SearchActivity),
                inputText.replace(" ", "").replace("\n", ""))
        }
    }

    override fun onSearchUserSuccess(searchUserBean: SearchUserBean) {
        page += 1
        mBinding.hint.text = ""
        mBinding.refreshLayout.finishRefresh()
        if (searchUserBean.has_next == 1) {
            mBinding.refreshLayout.finishLoadMore(true)
        } else {
            mBinding.refreshLayout.finishRefreshWithNoMoreData()
        }

        if (searchUserBean.page == 1) {
            mBinding.rv.adapter = searchUserAdapter
            mBinding.rv.scheduleLayoutAnimation()
            searchUserAdapter.setNewData(searchUserBean.body.list)
        } else {
            searchUserAdapter.addData(searchUserBean.body.list)
        }
        if (searchUserAdapter.data.size == 0) {
            mBinding.hint.text = "啊哦，没有数据"
        }
    }

    override fun onSearchUserError(msg: String?) {
        mBinding.refreshLayout.finishRefresh()
        if (page == 1) {
            if (searchPostAdapter.data.size != 0) {
                showToast(msg, ToastType.TYPE_ERROR)
            } else {
                mBinding.hint.text = msg
            }
            mBinding.refreshLayout.finishLoadMore()
        } else {
            mBinding.refreshLayout.finishLoadMore(false)
        }
    }

    override fun onSearchPostSuccess(searchPostBean: SearchPostBean) {
        page += 1
        mBinding.hint.text = ""
        mBinding.refreshLayout.finishRefresh()
        if (searchPostBean.has_next == 1) {
            mBinding.refreshLayout.finishLoadMore(true)
        } else {
            mBinding.refreshLayout.finishRefreshWithNoMoreData()
        }

        if (searchPostBean.page == 1) {
            mBinding.rv.adapter = searchPostAdapter
            mBinding.rv.scheduleLayoutAnimation()
            searchPostAdapter.addSearchPostData(searchPostBean.list, true)
        } else {
            searchPostAdapter.addSearchPostData(searchPostBean.list, false)
        }

        if (searchPostAdapter.data.size == 0) {
            mBinding.hint.text = "啊哦，没有数据"
        }
    }

    override fun onSearchPostError(msg: String?) {
        mBinding.refreshLayout.finishRefresh()
        if (page == 1) {
            if (searchPostAdapter.data.size != 0) {
                showToast(msg, ToastType.TYPE_ERROR)
            } else {
                mBinding.hint.text = msg
            }
            mBinding.refreshLayout.finishLoadMore()
        } else {
            mBinding.refreshLayout.finishLoadMore(false)
        }
    }

    override fun onKey(v: View?, keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_ENTER && event?.action == KeyEvent.ACTION_DOWN) {
            CommonUtil.hideSoftKeyboard(this@SearchActivity, v)
            mBinding.refreshLayout.autoRefresh(0, 300, 1f, false)
        }
        return false
    }

    override fun setStatusBar() {
        StatusBarUtil.setColor(this, ColorUtil.getAttrColor(this, R.attr.colorSurface), 0)
    }

    override fun afterTextChanged(s: Editable?) {
        val searchId = try { s.toString().toInt() } catch (e: Exception) { -1 }
        if (searchId != -1) {
            mBinding.suggestionText.apply {
                visibility = View.VISIBLE
                text = if (mBinding.btnByPost.isChecked) {
                    "检测到你输入了一串可能代表帖子id的数字，点击即可跳转到该帖子详情页"
                } else {
                    "检测到你输入了一串可能代表用户id的数字，点击即可跳转到该用户详情页"
                }
                setOnClickListener {
                    val intent = if (mBinding.btnByPost.isChecked) {
                        Intent(this@SearchActivity, NewPostDetailActivity::class.java).apply {
                            putExtra(Constant.IntentKey.TOPIC_ID, searchId)
                        }
                    } else {
                        Intent(this@SearchActivity, UserDetailActivity::class.java).apply {
                            putExtra(Constant.IntentKey.USER_ID, searchId)
                        }
                    }
                    startActivity(intent)
                }
            }
        } else {
            mBinding.suggestionText.visibility = View.GONE
        }
    }

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {

    }

}